### PR TITLE
Warn on unqualified top-level calls to Any or AnyRef methods

### LINKED ISSUE #17266: Warn when calling synchronized on Predef or on synthetic file objects
Consider the following code in `sync.scala`:

```scala
package myPackage:
	def test1 = synchronized { println("1 starting"); Thread.sleep(2000); println("Ok t1 finished.") }
	def test2 = this.synchronized { println("2 starting"); Thread.sleep(2000); println("Ok t2 finished"); }

object MyObject:
	def test3 = synchronized { println("3 starting"); Thread.sleep(2000); println("Ok t3 finished.") }
	def test4 = this.synchronized { println("4 starting"); Thread.sleep(2000); println("Ok t4 finished"); }
```

`scala-cli compile -Xprint:typer -Yprint-debug sync.scala` ouputs:

```scala
package <root>.this.myPackage {
    final lazy module val sync2$package: myPackage.this.sync2$package$ = 
      new myPackage.this.sync2$package$.<init>()
    final module class sync2$package$() extends Object.<init>() { 
      this: myPackage.this.sync2$package.type =>
      def test1: scala.this.Unit(inf) = 
        scala.this.Predef.synchronized[scala.this.Unit^(inf)](
          {
            scala.this.Predef.println("1 starting")
            java.this.lang.Thread.sleep(2000L)
            scala.this.Predef.println("Ok t1 finished.")
          }
        )
      def test2: scala.this.Unit(inf) = 
        this.synchronized[scala.this.Unit^(inf)](
          {
            scala.this.Predef.println("2 starting")
            java.this.lang.Thread.sleep(2000L)
            scala.this.Predef.println("Ok t2 finished")
          }
        )
    }
  }
  final lazy module val MyObject: <empty>.this.MyObject$ = 
    new <empty>.this.MyObject$.<init>()
  final module class MyObject$() extends Object.<init>() { 
    this: <empty>.this.MyObject.type =>
    def test3: scala.this.Unit(inf) = 
      MyObject$.this.synchronized[scala.this.Unit^(inf)](
        {
          scala.this.Predef.println("3 starting")
          java.this.lang.Thread.sleep(2000L)
          scala.this.Predef.println("Ok t3 finished.")
        }
      )
    def test4: scala.this.Unit(inf) = 
      this.synchronized[scala.this.Unit^(inf)](
        {
          scala.this.Predef.println("4 starting")
          java.this.lang.Thread.sleep(2000L)
          scala.this.Predef.println("Ok t4 finished")
        }
      )
  }
```

C.f., the first `synchronized` is desugared to `Predef.synchronized`. This sounds rather confusing.

We should probably issue a warning in this case, and also when calling `this.synchronized` from top-level functions.

More context in https://github.com/lampepfl/dotty/issues/7713#issuecomment-1508503085.

### PR DESCRIPTION
**Pull request by mbovel**
Fixes #17266

### CODE REVIEW

**Comment by mbovel** [comment]
We should probably extend that to `wait`, `notify` and maybe other `AnyRef` methods.

File: compiler/test/dotty/tools/dotc/CompilationTests.scala [hunk 1]
```diff
@@ -145,7 +145,7 @@ class CompilationTests {
       compileFilesInDir("tests/neg-custom-args/feature", defaultOptions.and("-Xfatal-warnings", "-feature")),
       compileFilesInDir("tests/neg-custom-args/no-experimental", defaultOptions.and("-Yno-experimental")),
       compileFilesInDir("tests/neg-custom-args/captures", defaultOptions.and("-language:experimental.captureChecking")),
-      compileFilesInDir("tests/neg-custom-args/explain", defaultOptions.and("-explain")),
+      compileFilesInDir("tests/neg-custom-args/explain", defaultOptions.and("-Xfatal-warnings", "-explain")),
```
[the remark this answers is not in the document]
Comment (mbovel) [reply compiler/test/dotty/tools/dotc/CompilationTests.scala:148]: Done ✅

**Comment by mbovel** [comment]
> We should probably extend that to wait, notify and maybe other AnyRef methods.

Done ✅

**Comment by som-snytt** [comment]
As motivation, I just did this on Scala 2!

That was using `eq` from a value class. It doesn't root import from `Predef`, but the class had an explicit import.

The unreliable Scala 2 equality warning gave me a heads up, but linting all such selections seems safer. I haven't checked if `eq` is covered here. This Scala 2 ticket is just about the message:

https://github.com/scala/bug/issues/12785

File: compiler/src/dotty/tools/dotc/reporting/messages.scala [hunk 2]
```diff
@@ -2284,6 +2284,25 @@ class PureExpressionInStatementPosition(stat: untpd.Tree, val exprOwner: Symbol)
         |It can be removed without changing the semantics of the program. This may indicate an error."""
 }
 
+class CallToAnyRefMethodOnPredef(stat: untpd.Tree, method: Symbol)(using Context)
+  extends Message(CallToAnyRefMethodOnPredefID) {
+  def kind = MessageKind.PotentialIssue
+  def msg(using Context) = i"Suspicious call to ${hl("Predef." + method.name)}"
+  def explain(using Context) =
+    i"""Top-level unqualified calls to ${hl("AnyRef")} or ${hl("Any")} methods such as ${hl(method.name.toString)} are
+       |resolved to calls on ${hl("Predef")}. This might not be what you intented."""
```
Comment (som-snytt) [inline compiler/src/dotty/tools/dotc/reporting/messages.scala:2293]: typo `intended`

File: compiler/src/dotty/tools/dotc/reporting/messages.scala [hunk 3]
```diff
@@ -2284,6 +2284,25 @@ class PureExpressionInStatementPosition(stat: untpd.Tree, val exprOwner: Symbol)
...
+}
+
+class CallToAnyRefMethodOnPackageObject(stat: untpd.Tree, method: Symbol)(using Context)
+  extends Message(CallToAnyRefMethodOnPackageObjectID) {
+  def kind = MessageKind.PotentialIssue
+  def msg(using Context) = i"Suspicious top-level call to ${hl("this." + method.name)}"
+  def explain(using Context) =
+    i"""Top-level calls to ${hl("AnyRef")} or ${hl("Any")} methods are resolved to calls on the
+       |synthetic package object generated for the current file. This might not be
+       |what you intented."""
```
Comment (som-snytt) [inline compiler/src/dotty/tools/dotc/reporting/messages.scala:2303]: same cut/paste

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala [hunk 4]
```diff
@@ -1090,6 +1090,17 @@ object RefChecks {
 
   end checkImplicitNotFoundAnnotation
 
+  def checkSynchronizedCall(tree: Tree, symbol: Symbol)(using Context) =
+    if symbol.exists && (symbol.owner == defn.ObjectClass || symbol.owner == defn.AnyClass) then
```
Comment (dwijnand) [inline compiler/src/dotty/tools/dotc/typer/RefChecks.scala:1094]: Perhaps we can add the last two stooges and go for
```suggestion
    if symbol.exists && defn.topClasses.contains(symbol.owner) then
```

**Review by dwijnand — APPROVED** [review]
LGTM.  Merge when you're happy to stop extending.

**Comment by odersky** [comment]
I'd like us to try a different warning. Predef is really not so special, after all. Consider this example:
```scala
object O

@main def foo =
  import O.*
  println(getClass)
```
This will print the `O$`, but it arguably should not. 

I think the warning should instead be triggered by this condition:

> When referring to a method from `Any` or `AnyRef` using a single identifier, one must be inside a named class or object (package objects don't count). 

Note that if we put the code in a class like this:
```scala
object O
object C:
  @main def foo =
    import O.*
    println(getClass)
```
we get a different warning, which is what we want:
```
6 |    println(getClass)
  |            ^^^^^^^^
  |            Reference to getClass is ambiguous.
  |            It is both defined in object C
  |            and imported subsequently by import O._
  |
  | longer explanation available when compiling with `-explain`
```

**Review by odersky — CHANGES_REQUESTED** [review]
By contrast see much less of an issue if we write `this.getClass` at the toplevel. Excluding this without excluding use of `this` in package objects as a whole looks unsystematic.

So either disallow all explicit uses of `this` in package objects or allow them all.

**Comment by mbovel** [comment]
@odersky I changed to condition to:

```scala
if tree.symbol.exists
   && defn.topClasses.contains(tree.symbol.owner)
   && (!ctx.owner.enclosingClass.exists || ctx.owner.enclosingClass.isPackageObject) then
  report.warning(UnqualifiedCallToAnyRefMethod(tree, tree.symbol), tree)
```

Seems to do what we want. What do you think?

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala [hunk 5]
```diff
@@ -1090,6 +1090,12 @@ object RefChecks {
 
   end checkImplicitNotFoundAnnotation
 
+  def checkSynchronizedCall(tree: Tree)(using Context) =
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/RefChecks.scala:1093]: Maybe rename that to `checkAnyMemberCall` or something like that?

**Review by odersky — APPROVED** [review]
Otherwise LGTM

**Pull request merged**